### PR TITLE
Independence number of graph and the maximum independent set of nodes of a graph.

### PR DESCRIPTION
**Pull request by NareshPeshwe**
Finds the independence number and iterator over all the independent set of nodes in the graph.

### CODE REVIEW

**Comment by MridulS** [comment]
@NareshPeshwe https://networkx.github.io/documentation/networkx-1.9.1/reference/generated/networkx.algorithms.approximation.independent_set.maximum_independent_set.html#networkx.algorithms.approximation.independent_set.maximum_independent_set

**Comment by MridulS** [comment]
This should give a better answer than the approximation but it is highly inefficient. What do you think @hagberg ?

File: networkx/algorithms/independence.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import networkx as nx
+from networkx.algorithms import *
```
Comment (MridulS) [inline networkx/algorithms/independence.py]: import the functions you are specifically using i.e. `enumerate_all_cliques` and `complement`

File: networkx/algorithms/independence.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+from networkx.utils import not_implemented_for
+__author__ = """Naresh Peshwe (nareshpeshwe12078@gmail.com)"""
```
Comment (MridulS) [inline networkx/algorithms/independence.py]: not needed

File: networkx/algorithms/independence.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+__all__ = ['enumerate_all_independentSets', 'independence_number']
+
+
+@not_implemented_for('directed')
+def enumerate_all_independentSets(G):
```
Comment (MridulS) [inline networkx/algorithms/independence.py]: Writing new functions for one liners should be avoided.

File: networkx/algorithms/independence.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	"""
+	return enumerate_all_cliques(complement(G))
+	
+
+def independence_number(G, independentsets=None):
+	""" Returns the independence number of an undirected graph
+
+	Parameters
+	----------
```
Comment (MridulS) [inline networkx/algorithms/independence.py]: description for `independentsets`?

**Comment by MridulS** [comment]
This code includes a mix a tabs and space. Have a look at https://www.python.org/dev/peps/pep-0008/ :)

**Comment by NareshPeshwe** [comment]
@MridulS Thanks for the remarks. Will make the changes. Yes i agree that this is inefficient. But as you have mentioned this would give a better answer than the approximation. And about the one liner function, should I combine both the functions into one and return a tuple ? Or is there any other way to go about it ?

**Comment by CrazyPython** [comment]
@NareshPeshwe I have reformatted your code and I have put the resulting file [here][https://gist.github.com/CrazyPython/60bf33fdcb33865af75109d8e1ed7b38).

- I replaced all the identation with spaces (PEP 8 convention)
- I renamed your method to use underscores instead of CamelCase (PEP 8 convention)
- I changed `... == None` to `... is None` (PEP 8 convention)

I think you should get a proper IDE - something like PyCharm. It can automatically reformat your code, rename functions/variables/classes/files, etc., detect bad formatting, and do autocomplete. My IDE highlighted all the problems noted above and fixed them for me too (after my agreement, of course).

If you didn't know, [PEP 8](https://gist.github.com/CrazyPython/60bf33fdcb33865af75109d8e1ed7b38) is the official Python document dictating standard formatting for python.

**Comment by CrazyPython** [comment]
@NareshPeshwe You also didn't use `import networkx as nx`. This again demonstrates the great power of a good IDE. (my IDE highlighted this)

**Comment by NareshPeshwe** [comment]
@CrazyPython Thank you for the remarks. It did save time :) .Yes, i didn't know about PEP 8 until @MridulS had pointed it out. I would use an appropriate IDE henceforth.

**Comment by CrazyPython** [comment]
@NareshPeshwe I hope you can make/copy the changes that you feel are appropriate from my gist. Also, by the way please capitalize your "i"'s. :)

File: networkx/algorithms/independence.py (hunk 1 above)
Comment (CrazyPython) [inline networkx/algorithms/independence.py]: using `from ... import *` is a bad practice. Some client IDEs may expose the unused functions imported by this statement, cluttering up the namespace of the client.

**Comment by ericmjl** [comment]
@CrazyPython @NareshPeshwe though PEP8 convention dictates lowercase for variable names, NetworkX (and network science) convention has been to call graphs by a single capitalized variable name. I will defer to @hagberg on what would be the best way forward for this specific little thing, though.

**Comment by NareshPeshwe** [comment]
@CrazyPython @ericmjl I too prefer using capitalized variable name to denote a graph. I went through the source code and at most of the occurrences "G" has been used. So I feel I should stick to it.

**Comment by CrazyPython** [comment]
@NareshPeshwe I agree. You haven't applied any of my other suggestions though.

**Comment by CrazyPython** [comment]
@NareshPeshwe From what I can tell, before you were either using Notepad/TextEdit or IDLE to do your editing. I sincerely hope you've switched.

**Comment by NareshPeshwe** [comment]
@CrazyPython Yes. I have started using PyCharm and its wonderful :) . I have applied the changes you suggested :) .

**Comment by CrazyPython** [comment]
@NareshPeshwe Great to hear! Always love to help out a fellow developer! BTW, you can press Alt+Enter to bring up the lightbulb/inspection menu.

**Comment by rossbar** [comment]
This PR was incorrectly closed automatically by GitHub when the default branch was renamed from master -> main. It looks like this PR has gone stale though (last activity ~2016). If there are plans to continue the work here, a new PR will have to be created. Please feel free to do so if there is interest in continuing this work - also feel free to ping for assistance, I'd be happy to help set up a new PR based on these commits.

**Pull request closed without merging**